src/eval.c

```diff
@@ -5394,6 +5466,805 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+		+ STRLEN(embedded[finished]->expr);
+
+	    mb_ptr_adv_nth(&p, embedding_start_length);  // the part of "${literal" or "$var"
+	    ++finished;
+	    continue; // the part of "}" (increasing)
+	}
+    }
+
+    return embedded;
+}
+
+    static int
+read_next_string_length(char_u *whole_string, int expect_literal_string, char_u *template)
+{
+    char_u	*p;
+    int		(*is_closing_quote)(char_u*);
+
+    // If entered with the closing quote of a template string into this function
+    if (whole_string[1] == NUL)
+    {
+	semsg(_("E1000: Unterminated template literal: %s"), template);
+	return -1;
+    }
+
+    p = whole_string + 1;  // Skip the head of '"' (or '\'')
+    is_closing_quote = expect_literal_string
+	? is_closing_single_quote
+	: is_closing_double_quote;
+
+    for (; !is_closing_quote(p); MB_PTR_ADV(p))
+    {
+	if (*p == NUL)
+	{
+	    semsg(_("E115: Missing quote: %s"), template);
+	    return -1;
+	}
+    }
+
+    return p - whole_string;
+}
+
+/*
+ * Get a first "expr" of "${" "expr" "}" from whole_embedding.
+ *
+ * whole_embedding starts with "${": e.g. "${10}'", "${var} Hello :D'".
+ *
+ * The result MUST BE EXECUTED BY free() later! (if it is not NULL.)
+ */
+    static embedded_t*
+read_next_embedded_lit(char_u *whole_embedding, char_u *template)
+{
+    const size_t	LITERAL_EMBEDDING_START_LENGTH = 2;  // "${"
+    char_u*		embedded_start = whole_embedding + LITERAL_EMBEDDING_START_LENGTH;
+    size_t		expr_length = 0;  // "${" expr_length "}"
+    size_t		nested_blocks = 0;  // {}'s nests: ${{'x': 10}}, ${something({})}
+    char_u*		p;
+
+    for (p = embedded_start; !(nested_blocks == 0 && *p == '}'); MB_PTR_ADV(p))
+    {
+	switch (*p)
+	{
+	    case '{':
+		++nested_blocks;
+		break;
+	    case '}':
+		--nested_blocks;
+		break;
+	    case '"':
+	    case '\'':
+	    {
+		int length = read_next_string_length(p, (*p == '\''), template);
+		if (length == -1)
+		{
+		    return NULL;
+		}
+		expr_length += length;
+		mb_ptr_adv_nth(&p, length);
+		break;
+	    }
+	    case NUL:
+		semsg(_("E1000: Unterminated template literal: %s"), template);
+		return NULL;  // failed! a closing "}" couldn't be found.
+	}
+	++expr_length;
+    }
+
+    return new_embedded(TRUE, embedded_start, expr_length);
+}
+
+/*
+ * Get a first "var" of "$" "var" from whole_embedding.
+ *
+ * whole_embedding starts with "$": e.g. "$var`", "$v_i_m".
+ * (like "$x[0]" and "$x.y" only parses the part of "$x".)
+ *
+ * The result MUST BE EXECUTED BY free() later! (if it is not NULL.)
+ */
+    static embedded_t*
+read_next_embedded_var(char_u *whole_embedding, char_u *template)
+{
+    const size_t	NUM_LAST_NON_SNEAK_CHAR = 1;
+    char_u*		expr_start = whole_embedding + 1;  // skip a head '$'
+    char_u*		p;
+
+    if (!is_sneak_case_head_char(expr_start[0]))
+    {
+	semsg(_("E1000: Missing variable name on a $: %s"), template);
+	return NULL;
+    }
+
+    for (p = expr_start + 1; *p != NUL; MB_PTR_ADV(p))
+    {
+	if (!is_sneak_case_char(*p))
```

Comment:
I wonder if we really need this function.
Isn't `!ASCII_ISALNUM(*p) && *p != '_'` enough?

Reply:
Exactly xD
I fixed!